pkg/api/validate/dynamic/dynamic.go

```diff
@@ -617,34 +623,92 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 			)
 		}
 
-		if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
-			if ss.SubnetPropertiesFormat != nil &&
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup != nil {
-				expectedNsgID, err := subnet.NetworkSecurityGroupID(oc, s.ID)
+		subnetByID[s.ID] = ss
+	}
+	return subnetByID, nil
+}
+
+// checkByoNSG checks whether all the subnets have or don't have NSG attached.
+// when the BYONsg feature flag is on and only some of the subnets are attached with an NSG,
+// it returns an error.  If none of the subnets is attached, it's no longer BYO NSG and the
+// cluster installation process should fall back to using the managed nsg.
+func (dv *dynamic) checkByoNSG(subnetByID map[string]*mgmtnetwork.Subnet) (api.PreconfiguredNSG, error) {
+	var attached int
+	for _, subnet := range subnetByID {
+		if subnetHasNSGAttached(subnet) {
+			attached++
+		}
+	}
+
+	// all subnets have an attached NSG
+	if attached == len(subnetByID) {
+		dv.log.Info("all subnets are attached, BYO NSG")
+		return api.PreconfiguredNSGEnabled, nil // correct setup by customer
+	}
+
+	// no subnets have attached NSG, fallback
+	if attached == 0 {
+		dv.log.Info("no subnets are attached, no longer BYO NSG. Fall back to using cluster NSG.")
+		return api.PreconfiguredNSGDisabled, nil
+	}
+
+	// some subnets have NSGs attached, error out
+	return api.PreconfiguredNSGDisabled,
+		&api.CloudError{
+			StatusCode: http.StatusBadRequest,
+			CloudErrorBody: &api.CloudErrorBody{
+				Code:    api.CloudErrorCodeInvalidLinkedVNet,
+				Message: errMsgNSGNotProperlyAttached,
+			},
+		}
+}
+
+func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster, subnets []Subnet) error {
+	subnetByID, err := dv.createSubnetMapByID(ctx, subnets)
+	if err != nil {
+		return err
+	}
+
+	if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
+		if oc.Properties.NetworkProfile.PreconfiguredNSG == api.PreconfiguredNSGEnabled {
+			oc.Properties.NetworkProfile.PreconfiguredNSG, err = dv.checkByoNSG(subnetByID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// we're parsing through the subnets slice, not the map because we'll return consistent error messages on creation
+	for _, s := range subnets {
+		ss := subnetByID[s.ID]
+
+		switch oc.Properties.ProvisioningState {
+		case api.ProvisioningStateCreating:
```

Comment:
nit: it feels strange to define a `switch` block for a single `case`. `if/else` reads a bit more like natural english here.